File: OpenBullet2.Web/Dtos/Proxy/ProxyFiltersDto.cs
using OpenBullet2.Web.Dtos.Common;
using RuriLib.Models.Proxies;

namespace OpenBullet2.Web.Dtos.Proxy;

/// <summary>
/// Filters to describe a subset of proxies.
/// </summary>
public class ProxyFiltersDto : PaginationDto
{
    /// <summary>
    /// The id of the proxy group. Set to -1 for all, where supported.
    /// </summary>
    public int ProxyGroupId { get; set; } = -1;

    /// <summary>
    /// The search term to filter results by the proxy host,
    /// port, username or country. Optional.
    /// </summary>
    public string? SearchTerm { get; set; } = null;

    /// <summary>
    /// The proxy type filter, if any.
    /// </summary>
    public ProxyType? Type { get; set; } = null;

    /// <summary>
    /// The proxy status filter, if any.
    /// </summary>
    public ProxyWorkingStatus? Status { get; set; } = null;
    
    /// <summary>
    /// The field to sort proxies by, if any.
    /// </summary>
    public ProxySortField? SortBy { get; set; } = null;
    
    /// <summary>
    /// Whether to sort the proxies in descending order.
    /// </summary>
    public bool SortDescending { get; set; } = false;
}

File: openbullet2-web-client/src/app/main/dtos/proxy/proxy-filters.dto.ts
import { ProxyQuality } from '../../enums/proxy-quality';
import { ProxyType } from '../../enums/proxy-type';
import { ProxyWorkingStatus } from '../../enums/proxy-working-status';

export enum ProxySortField {
  Host = 'host',
  Port = 'port',
  Username = 'username',
  Password = 'password',
  Country = 'country',
  Ping = 'ping',
  LastChecked = 'lastChecked',
}

export interface ProxyFiltersDto {
  pageNumber: number | null;
  pageSize: number | null;
  proxyGroupId: number;
  searchTerm: string | null;
  type: ProxyType | null;
  status: ProxyWorkingStatus | null;
  quality: ProxyQuality | null;
  sortBy: ProxySortField | null;
  sortDescending: boolean;
}
